sssd_test_framework/utils/sbus.py

```diff
@@ -0,0 +1,672 @@
+            delim = ","
+        return val
+
+    @classmethod
+    def parse(cls, result: DBUSResult) -> DBUSTypeArray:
+        # dbus-send doesn't return "dict {...}" but "array [ dict entry (...) ... ]"
+        dict = None
+        old = result.text
+        try:
+            dict = DBUSTypeDict.parse(result)
+        except TypeError:
+            pass
+        if dict is not None:
+            result.text = old
+            raise TypeError("Not an array")
+
+        res = re.match(r"^ *array +\[", result.text)
+        if res is None:
+            raise TypeError("Not an array")
+
+        array = []
+        result.text = result.text.removeprefix(res.group(0)).strip()
+        res = re.match(r"^ *\] *", result.text)
+        while len(result.text) > 0 and res is None:
+            value = cls.get_python_value(result)
+            array.append(value)
+            result.text = result.text.strip()
+            res = re.match(r"^\] *", result.text)
+        if res is None:
+            raise TypeError("Malformed an array")
+
+        result.text = result.text.removeprefix(res.group(0))
+        return DBUSTypeArray(array)
+
+
+class DBUSTypeDict(DBUSType):
+    def __init__(self, value: dict):
+        self.value = value
+        """The actual value stored as a python type"""
+
+    def build(self) -> str:
+        k = list(self.value)[0]
+        val = f"dict:{self.Types.get_dbus_type(k)}:{self.Types.get_dbus_type(self.value[k])}:"
+        delim = ""
+        for k, v in self.value.items():
+            val += f"{delim}{str(k)},{str(v)}"
+            delim = ","
+        return val
+
+    @classmethod
+    def _parse_dict_entry(cls, result: DBUSResult) -> tuple:
+        res = re.match(r"^ *dict entry *\(", result.text)
+        if res is None:
+            raise TypeError("Not a dictionary entry")
+
+        result.text = result.text.removeprefix(res.group(0))
+        key = cls.get_python_value(result)
+        value = cls.get_python_value(result)
+
+        res = re.match(r"^ *\) *", result.text)
+        if res is None:
+            raise TypeError("Malformed dictionary entry")
+
+        result.text = result.text.removeprefix(res.group(0)).strip()
+        return key, value
+
+    @classmethod
+    def parse(cls, result: DBUSResult) -> DBUSTypeDict:
+        res = re.match(r"^ *array +\[ *(dict entry *\(.+\)) *\] *", result.text)
+        if res is None:
+            raise TypeError("Not a dictionary")
+
+        res2 = DBUSResult(res.group(1).strip())
+        dict = {}
+        while len(res2.text) > 0:
+            key, value = cls._parse_dict_entry(res2)
+            dict[key] = value.value if isinstance(value, DBUSTypeVariant) else value
+
+        result.text = result.text[res.span()[1] :].strip()
+        return DBUSTypeDict(dict)
+
+
+class DBUSObject(MultihostUtility[MultihostHost]):
+    """
+    D-Bus access.
+
+    Objects of this class represent the root or child D-Bus nodes.
+    """
+
+    class ProxyObject(ABC):
```

Comment:
It would be good to avoid class nesting in this case. It makes the code harder to read and there is no use for it.